File: acoustic_simulator/tests/test_source_mixer_update_geometry.m
function tests = test_source_mixer_update_geometry()
%TEST_SOURCE_MIXER_UPDATE_GEOMETRY  Tests for live geometry updates.
%
%   The GUI scene sliders rebuild geometry on every change and call
%   SourceMixer.update_geometry(new_geo) so the next mix() block picks
%   up the new gains/delays. These tests guarantee:
%     * gains and frac-delays are actually swapped in
%     * per-source history rings are resized correctly
%     * a step change in geometry does NOT click on the boundary
%       (the mixer produces a finite, bounded output)
    tests = functiontests(localfunctions);
end

function setupOnce(tc) %#ok<*DEFNU>
    here = fileparts(fileparts(mfilename('fullpath')));
    addpath(fullfile(here, 'config'));
    addpath(fullfile(here, 'core'));
    tc.TestData.cfg = default();
end

function test_update_changes_gains(tc)
    cfg  = tc.TestData.cfg;
    geo1 = build_geometry(cfg);
    mix  = SourceMixer(cfg, geo1);

    cfg2 = cfg;
    cfg2.slant_dist = 6.0;                  % human further → smaller speech gain
    geo2 = build_geometry(cfg2);
    mix.update_geometry(geo2);

    g_old = geo1.gains_speech(cfg.mwf.ref_mic);
    g_new = geo2.gains_speech(cfg.mwf.ref_mic);
    verifyLessThan(tc, g_new, g_old, ...
        'Speech gain must drop when the human moves farther away.');

    % And the mixer must actually use the new value on the next block.
    N   = cfg.frame_size;
    s   = randn(N, 1);
    out = mix.mix(s, zeros(N,1), zeros(N,1));
    [~, m0] = min(geo2.frac_delays_speech);
    ratio   = max(abs(out(:, m0))) / max(abs(s));
    verifyEqual(tc, ratio, g_new, 'AbsTol', 1e-6);
end

function test_update_resizes_history_without_click(tc)
%TEST_UPDATE_RESIZES_HISTORY_WITHOUT_CLICK  Run a steady-state block,
%   step the geometry, then run another block of the same input — the
%   second block must be finite and free of NaN/Inf regardless of the
%   history-buffer resize.
    cfg = tc.TestData.cfg;
    geo = build_geometry(cfg);
    mix = SourceMixer(cfg, geo);

    N = cfg.frame_size;
    s = sin(2*pi*250*(0:N-1).'/cfg.fs);

    mix.mix(s, zeros(N,1), zeros(N,1));     % prime history

    cfg2 = cfg;
    cfg2.slant_dist = 0.5;                  % much closer → tiny delays
    mix.update_geometry(build_geometry(cfg2));
    out_small = mix.mix(s, zeros(N,1), zeros(N,1));
    verifyTrue(tc, all(isfinite(out_small(:))), ...
        'Mixer output contains NaN/Inf after shrinking history.');
    verifyLessThanOrEqual(tc, max(abs(out_small(:))), 2, ...
        'Mixer output blew up after shrinking history.');

    cfg3 = cfg;
    cfg3.slant_dist = 8.5;                  % much farther → large delays
    mix.update_geometry(build_geometry(cfg3));
    out_big = mix.mix(s, zeros(N,1), zeros(N,1));
    verifyTrue(tc, all(isfinite(out_big(:))), ...
        'Mixer output contains NaN/Inf after growing history.');
    verifyLessThanOrEqual(tc, max(abs(out_big(:))), 2, ...
        'Mixer output blew up after growing history.');
end

function test_update_rejects_n_mic_change(tc)
%TEST_UPDATE_REJECTS_N_MIC_CHANGE  Changing cfg.n_mics at runtime would
%   break downstream buffer shapes, so update_geometry must throw.
    cfg = tc.TestData.cfg;
    mix = SourceMixer(cfg, build_geometry(cfg));
    cfg2 = cfg;
    cfg2.n_mics = cfg.n_mics + 2;
    geo2 = build_geometry(cfg2);
    f = @() mix.update_geometry(geo2);
    verifyError(tc, f, 'SourceMixer:update_geometry:NMicChange');
end

File: simulator/acoustic_simulator_py/tests/test_source_mixer_update_geometry.py
"""Tests for :meth:`SourceMixer.update_geometry`.

Mirrors ``tests/test_source_mixer_update_geometry.m``. The GUI scene
sliders call this method on every drag; we guarantee that the gains and
fractional delays actually swap in, the per-source history rings resize
without clicking, and a mic-count change is rejected loudly.
"""

from __future__ import annotations

import numpy as np
import pytest

from backend.config import default
from backend.core import NMicChangeError, SourceMixer, build_geometry


def test_update_changes_gains() -> None:
    cfg = default()
    geo1 = build_geometry(cfg)
    mixer = SourceMixer(cfg, geo1)

    cfg2 = default()
    cfg2.slant_dist = 6.0    # human further → smaller speech gain
    geo2 = build_geometry(cfg2)
    mixer.update_geometry(geo2)

    ref0 = cfg.mwf.ref_mic - 1
    g_old = float(geo1.gains_speech[ref0])
    g_new = float(geo2.gains_speech[ref0])
    assert g_new < g_old, "speech gain must drop when the speaker moves farther"

    # Next block must actually use the new gain.
    n = cfg.frame_size
    rng = np.random.default_rng(1)
    src = rng.standard_normal(n)
    out = mixer.mix(src, np.zeros(n), np.zeros(n))
    m0 = int(np.argmin(geo2.frac_delays_speech))
    ratio = float(np.max(np.abs(out[:, m0])) / np.max(np.abs(src)))
    assert ratio == pytest.approx(g_new, abs=1e-6)


def test_update_resizes_history_without_click() -> None:
    """Prime the history, then shrink (slant 0.5 m) and grow (slant 8.5 m)
    the geometry. Output stays finite and bounded across the boundary."""
    cfg = default()
    mixer = SourceMixer(cfg, build_geometry(cfg))
    n = cfg.frame_size
    s = np.sin(2 * np.pi * 250 * np.arange(n) / cfg.fs)
    mixer.mix(s, np.zeros(n), np.zeros(n))    # prime

    cfg_close = default()
    cfg_close.slant_dist = 0.5
    mixer.update_geometry(build_geometry(cfg_close))
    out_small = mixer.mix(s, np.zeros(n), np.zeros(n))
    assert np.all(np.isfinite(out_small)), "history-shrink produced NaN/Inf"
    assert float(np.max(np.abs(out_small))) <= 2.0, "history-shrink blew up"

    cfg_far = default()
    cfg_far.slant_dist = 8.5
    mixer.update_geometry(build_geometry(cfg_far))
    out_big = mixer.mix(s, np.zeros(n), np.zeros(n))
    assert np.all(np.isfinite(out_big)), "history-grow produced NaN/Inf"
    assert float(np.max(np.abs(out_big))) <= 2.0, "history-grow blew up"


def test_update_rejects_n_mic_change() -> None:
    cfg = default()
    mixer = SourceMixer(cfg, build_geometry(cfg))
    cfg2 = default()
    cfg2.n_mics = cfg.n_mics + 2
    geo2 = build_geometry(cfg2)
    with pytest.raises(NMicChangeError):
        mixer.update_geometry(geo2)
